src/main/scala/wdl4s/Task.scala

```diff
@@ -30,7 +31,87 @@ object Task {
       case x: Ast => ParameterCommandPart(x, wdlSyntaxErrorFormatter)
     }
 
-    val outputs = ast.findAsts(AstNodeName.Output) map { TaskOutput(_, declarations, wdlSyntaxErrorFormatter) }
+    val variablesReferencedInCommand = for {
+      param <- commandTemplate collect { case x: ParameterCommandPart => x }
+      variable <- param.expression.variableReferences
+    } yield variable
+
+    variablesReferencedInCommand foreach { variable =>
+      if (!declarations.map(_.name).contains(variable.getSourceString)) {
+        throw new SyntaxError(wdlSyntaxErrorFormatter.commandExpressionContainsInvalidVariableReference(taskNameTerminal, variable))
+      }
+    }
+
+    /**
+      * Ensures that the current declaration doesn't have a name conflict with another declaration
+      * and that the expression for the current declaration only has valid variable references in it
+      *
+      * @param accumulated The declarations that come lexically before 'current'
+      * @param current The declaration being validated
+      */
+    def validateDeclaration(accumulated: Seq[Declaration], current: Declaration): Seq[Declaration] = {
```

Comment:
and the way this is used will also result in fast failing on the first bad declaration